lib/gemini.c

```diff
@@ -0,0 +1,296 @@
+
+#include <string.h>
+#include "urldata.h"
+#include "transfer.h"
+#include "connect.h"
+#include "cfilters.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "select.h"
+#include "strdup.h"
+#include "url.h"
+#include "curl_printf.h"
+#include "gemini.h"
+#include "memdebug.h"
+
+static CURLcode gemini_setup_connection(struct Curl_easy *data,
+                                        struct connectdata *conn)
+{
+  (void)conn;
+  DEBUGASSERT(data->req.p.gemini == NULL);
+
+  data->req.p.gemini = calloc(1, sizeof(struct GEMINI));
+
+  if(!data->req.p.gemini)
+    return CURLE_OUT_OF_MEMORY;
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_connect(struct Curl_easy *data, bool *done)
+{
+  char *url = data->state.url;
+  struct GEMINI *gemini = data->req.p.gemini;
+
+  /* url is guaranteed to be non-NULL and non-empty */
+  DEBUGASSERT(url);
+
+  /* gemini is guaranteed to be non-NULL */
+  DEBUGASSERT(gemini);
+
+  /* url must not start with 0xFEFF */
+  if((unsigned char)url[0] == 0xFF && (unsigned char)url[1] == 0xFE) {
+    failf(data, "URL starts with 0xFEFF");
+    return CURLE_URL_MALFORMAT;
+  }
+
+  /* url must be 1024 bytes or less */
+  if(strlen(url) > 1024) {
+    failf(data, "URL is too long");
+    return CURLE_URL_MALFORMAT;
+  }
+
+  /* Copy the URL and termination to the GEMINI struct */
+  gemini->reqindex = 0;
+  gemini->reqlen = curl_msnprintf(gemini->request,
+                                  sizeof(gemini->request),
+                                  "%s%s",
+                                  url,
+                                  GEMINI_TERMINATION);
+
+  if(gemini->reqlen < 0) {
+    failf(data, "Failed to copy URL to GEMINI struct");
+    return CURLE_URL_MALFORMAT;
+  }
+
+  *done = true;
+  return CURLE_OK;
+}
+
+static CURLcode gemini_connecting(struct Curl_easy *data, bool *done)
+{
+  struct connectdata *conn = data->conn;
+  CURLcode result;
+
+  result = Curl_conn_connect(data, FIRSTSOCKET, TRUE, done);
+  if(result)
+    connclose(conn, "Failed TLS connection");
+
+  *done = TRUE;
+  return result;
+}
+
+static CURLcode gemini_doing(struct Curl_easy *data, bool *done)
+{
+  CURLcode result = CURLE_OK;
+  struct connectdata *conn = data->conn;
+  curl_socket_t sockfd = conn->sock[FIRSTSOCKET];
+  struct GEMINI *gemini = data->req.p.gemini;
+  ssize_t rx_size = GEMINI_STATUS_SIZE + 1 + GEMINI_META_MAX;
+
+  /* Send request */
+  if(gemini->reqindex < gemini->reqlen) {
+    result = Curl_write(data,
+                        sockfd,
+                        &gemini->request[gemini->reqindex],
+                        gemini->reqlen - gemini->reqindex,
+                        &data->info.request_size);
+
+    if(result)
+      return result;
+
+    gemini->reqindex += data->info.request_size;
+
+    /* If we haven't sent the entire request yet, return */
+    if(gemini->reqindex < gemini->reqlen)
+      return CURLE_OK;
+  }
+
+  /* Receive response header */
```

Comment:
in the `doing` function, libcurl should be sending the request, it is not for receiving content. That's the PERFORM state.